stablehlo/dialect/StablehloBytecode.cpp

```diff
@@ -0,0 +1,695 @@
+/* Copyright 2022 The StableHLO Authors.
+StablehloBytecode.cpp - StableHLO Bytecode Implementation */
+
+#include "stablehlo/dialect/StablehloBytecode.h"
+
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Bytecode/BytecodeImplementation.h"
+#include "mlir/IR/Diagnostics.h"
+#include "stablehlo/dialect/StablehloOps.h"
+
+//===----------------------------------------------------------------------===//
+// Debug Trace Helpers
+//===----------------------------------------------------------------------===//
+
+// Enable logging with flag:
+//   stablehlo-opt -debug-only=stablehlo-bytecode [...]
+//
+// Extract after function name, remove namespace.
+//   Called: write(mlir::stablehlo::TokenType, mlir::DialectBytecodeWriter ...
+//   ***Not Implemened: write(...
+#define _EXTRACT_AFTER(a, b) \
+  llvm::StringRef(a).substr(llvm::StringRef(a).find(b))
+
+#define _LOG_CALL_TO(func)                                                    \
+  DEBUG_WITH_TYPE(                                                            \
+      "stablehlo-bytecode",                                                   \
+      llvm::errs() << "Called: " << _EXTRACT_AFTER(__PRETTY_FUNCTION__, func) \
+                   << '\n')
+
+#define LOG_WRITE_CALL _LOG_CALL_TO("write")
+#define LOG_READ_CALL _LOG_CALL_TO(__func__)
+#define LOG_NOT_IMPLEMENTED \
+  DEBUG_WITH_TYPE(          \
+      "stablehlo-bytecode", \
+      llvm::errs() << "***Not Implemented: " << __PRETTY_FUNCTION__ << '\n')
```

Comment:
It's helpful to have these logging facilities to better understand what's going on! In the future, we'll likely want to standardize on our approach to logging as part of #15. Meanwhile this LGTM.

Reply:
Agree - not sure if we want to make our own small logging library built on top of LLVM to help conform to our standards once they exist, but may be helpful.